platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,231 @@
+     * @param writableStates the readable states
+     */
+    public WritableRosterStore(@NonNull final WritableStates writableStates) {
+        this.writableStates = writableStates;
+        this.rosterStateAccessor = new ReadableRosterStore(writableStates);
+        this.rosterState = writableStates.getSingleton(RosterStateId.ROSTER_STATES_KEY);
+        this.rosterMap = writableStates.get(RosterStateId.ROSTER_KEY);
+    }
+
+    /**
+     * Set the candidate roster if valid and doesn't yet exist in the state.
+     *
+     * @param candidateRoster a candidate roster to set.
+     */
+    @Override
+    public void setCandidateRoster(@NonNull final Roster candidateRoster) {
+        Objects.requireNonNull(candidateRoster);
+        RosterValidator.validate(candidateRoster);
+
+        // update the roster state
+        final RosterState currentRosterState = rosterStateOrThrow();
+        final Bytes candidateRosterHash = RosterUtils.hash(candidateRoster).getBytes();
+        final Builder rosterStateBuilder = RosterState.newBuilder()
+                .candidateRosterHash(candidateRosterHash)
+                .roundRosterPairs(currentRosterState.roundRosterPairs());
+        this.rosterState.put(rosterStateBuilder.build());
+
+        // update the roster map and commit the changes
+        this.rosterMap.put(ProtoBytes.newBuilder().value(candidateRosterHash).build(), candidateRoster);
+        commit();
+    }
+
+    /**
+     * Determines the initial active roster based on the given software version and initial state.
+     * The active roster is obtained by adopting the candidate roster if a software upgrade is detected.
+     * Otherwise, the active roster is retrieved from the state or an exception is thrown if not found.
+     *
+     * @param version the software version of the current node
+     * @param initialState the initial state of the platform
+     * @return the active roster which will be used by the platform
+     */
+    @NonNull
+    @Override
+    public Roster determineActiveRoster(
+            @NonNull final SoftwareVersion version, @NonNull final ReservedSignedState initialState) {
+        final boolean softwareUpgrade = detectSoftwareUpgrade(version, initialState.get());
+
+        if (!softwareUpgrade) {
+            final Roster lastUsedActiveRoster = rosterStateAccessor.getActiveRoster();
+            // not in software upgrade mode (i.e., normal restart), return the
+            // active roster present in the state or throw if not found.
+            return Objects.requireNonNull(
+                    lastUsedActiveRoster, "Active Roster must be present in the state during normal network restart.");
+        }
+
+        // software upgrade is detected, we adopt the candidate roster
+        final Roster candidateRoster = Objects.requireNonNull(
+                rosterStateAccessor.getCandidateRoster(),
+                "Candidate Roster must be present in the state before attempting to adopt a roster.");
```

Comment:
Do we actually want to throw in this case? It seems a valid use case to perform an upgrade, having *not* set a new candidate roster. I would assume in that case that we just continue using the current active roster

Reply:
yes, we do. There are other ways to provide an active roster than a call to this method such as via a `GenesisRoster` or an `OverrideRoster`. In the case there's no existing `candidateRoster` in the state and an upgrade is required, an `OverrideRoster` will have to be provided. Implementation of that is pending though.

Reply:
> an OverrideRoster will have to be provided

Assuming an upgrade is taking place, where no roster changes are required, and thus no candidate roster has been set: why should we force the presence of an `OverrideRoster`, instead of just continuing to use the roster that already exists?

Reply:
In that assumed case, this method will not be called. See https://github.com/hashgraph/hedera-services/pull/15442#discussion_r1786180225

Reply:
I think I understand your explanation, but I don't think it is the right solution.

It doesn't make sense to force the caller to provide an "override roster" in the case where they simply want to continue operating with the existing active roster.

Reply:
I agree with @litt3.  We should be able to do software upgrades without a candidate or override roster set.  (See config-only software upgrades).